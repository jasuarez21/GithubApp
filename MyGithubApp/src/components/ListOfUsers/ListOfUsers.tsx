import React, {useEffect} from 'react';
import {widthPercentageToDP as wp, heightPercentageToDP as hp} from 'react-native-responsive-screen';
import {useSelector} from 'react-redux';
import {StackScreenProps} from '@react-navigation/stack';
import Header from '../Header/Header';
import {TouchableHighlight, View, Text, ScrollView,  StyleSheet, Image} from 'react-native';
import { User } from '../../interfaces/User';


interface Props extends StackScreenProps<any, any> {}

const List = ({navigation}: Props) => {
    const user: [User] = useSelector((store: any) => store.user);
    let usersList: [User] = user;
    useEffect(() => {
        usersList = user
    }, [user])
    return ( 
        <View style={styles.backgroundView}>
        <Header />
        <Text style={styles.title}>Lista de usuarios</Text>
            <ScrollView style={styles.scrollContainer}>
            {
                    usersList.map((elem: User) => (
                            <TouchableHighlight style={styles.userTarget} onPress={() => navigation.navigate('DetailOfUser', {userSelected: elem})}>
                                <>
                                    <Image
                                        testID = {elem?.login}
                                        style={styles.photo}
                                        source={{
                                            uri: elem?.avatar_url,
                                        }}
                                    />
                                        <Text style={styles.username}>{elem?.login}</Text>
                                </>
                            </TouchableHighlight>
                    ))
            }
            {
                usersList[0]?.login === undefined ? (
                    <Text style={styles.errorMessage}>No hay resultados</Text>
                ):(
                    <Text style={styles.findUser}>Se han encontrado {usersList.length} resultados</Text>
                )
            }  
            </ScrollView>
        </View>  
    )
}

const styles = StyleSheet.create({
    scrollContainer: {
        width: 'auto',
        height: 'auto',
        marginTop: wp('3%')
    },
    backgroundView: {
        height: 'auto',
    },
    errorMessage:{
        fontSize: hp('2%'),
        color: '#03045e',
        textAlign: 'center'
    },
    findUser: {
        fontSize: hp('1.4%'),
        color: '#03045e',
        position: 'absolute',
        marginLeft: wp('50%')
    },
    title: {
        fontSize: hp('2%'),
        textTransform: 'uppercase',
        marginTop: hp('6%'),
        marginLeft: wp('6%'),
        color: '#03045e',
        borderBottomWidth: 1,
        width: wp('85%'),
        borderBottomColor: '#03045e'
    },
    photo: {
        height: hp('12%'),
        width: wp('20%'),
        borderRadius: 50,
        marginTop: hp('2%')
    },
    username: {
        fontSize: hp('2%'),
        marginTop: hp('6%'),
        textTransform: 'uppercase',
        color: '#03045e'
    },
    userTarget: {
        height: hp('17%'),
        width: wp('85%'),
        marginLeft: wp('6%'),
        marginTop: hp('2.5%'),
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-around',
        borderWidth: 1,
        borderRadius: 25,
        borderColor: '#023e8a',
        backgroundColor: '#caf0f8'
    },
    erroMessage: {
        color: 'red',
        fontSize: hp('2.2%')
    }
})

export default List;