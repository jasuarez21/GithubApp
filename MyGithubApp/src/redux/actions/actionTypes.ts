export default {
    GET_USER: 'GET_USER',
    ERROR: 'ERROR'
}