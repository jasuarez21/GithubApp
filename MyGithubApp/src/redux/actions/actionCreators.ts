import axios from 'axios';
import actionTypes from './actionTypes';
import {Dispatch} from 'redux';

export function getUsers(user: string){
    console.log(user)
    return async (dispatch: Dispatch) => {
        try{
            const {data} = await axios.get('the url')
            dispatch({
                type: actionTypes.GET_USER,
                data,
            })
        } catch (error) {
            dispatch({
                type: actionTypes.ERROR
            })
        }
    }
}