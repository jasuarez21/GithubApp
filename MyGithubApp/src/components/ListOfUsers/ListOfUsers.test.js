import { render, fireEvent } from '@testing-library/react-native';
import React from 'react';
import ConfigureStore from 'redux-mock-store';
import * as actions from '../../redux/actions/actionCreators';
import { Provider } from 'react-redux';
import ListOfUsers from './ListOfUsers';
import actionTypes from '../../redux/actions/actionTypes';

jest.mock('../../redux/actions/actionCreators');

describe('When invoked a ListOfUsers component', () => {
    const mockStore = ConfigureStore();
    const store = mockStore({
        user : [{
            login: 'Josep'
        },
        {
            login: 'Juanito' 
        }]
    })
    let route;
    let navigation;
    beforeEach(() => {
        navigation = {
            navigate: jest.fn()
        }
        dispatch = jest.fn();
        route = {
            params: {
                userSelected : 'Josep'
            }
        }
    })
    test('Should render the List of Users page', () => {
        const list = render(<Provider store={store}><ListOfUsers navigation={navigation} route={route} /></Provider>)
        expect(list).toMatchSnapshot()
    })
    test('Should navigate to detail of user when i click a target', async () => {
        const { getByTestId } = render(<Provider store={store}><ListOfUsers navigation={navigation} route={route} /></Provider>);
        const detailUser = getByTestId('Josep');
        await fireEvent.press(detailUser);
        expect(navigation.navigate).toHaveBeenCalled();
    })
})