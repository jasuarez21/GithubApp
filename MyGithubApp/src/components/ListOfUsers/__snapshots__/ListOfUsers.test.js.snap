// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When invoked a ListOfUsers component Should render the List of Users page 1`] = `
<View
  style={
    Object {
      "height": "auto",
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "display": "flex",
        "flexDirection": "row",
        "marginTop": 15,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://i.postimg.cc/W4pH9FN9/icons8-github-50.png",
        }
      }
      style={
        Object {
          "height": 50,
          "marginLeft": -40,
          "marginTop": 10,
          "width": 50,
        }
      }
    />
    <TextInput
      autoCapitalize="none"
      defaultValue=""
      onChangeText={[Function]}
      placeholder="Introduce el usuario"
      style={
        Object {
          "backgroundColor": "#caf0f8",
          "borderColor": "#023e8a",
          "borderRadius": 25,
          "borderWidth": 1,
          "height": 50,
          "marginLeft": 20,
          "marginTop": 10,
          "padding": 10,
          "width": 250,
        }
      }
      testID="searchInput"
      value=""
    />
  </View>
  <Text
    style={
      Object {
        "borderBottomColor": "#03045e",
        "borderBottomWidth": 1,
        "color": "#03045e",
        "fontSize": 18,
        "marginLeft": 30,
        "marginTop": 50,
        "width": 400,
      }
    }
  >
    Lista de usuarios
  </Text>
  <RCTScrollView
    style={
      Object {
        "height": "auto",
        "marginTop": 20,
        "width": 500,
      }
    }
  >
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#caf0f8",
            "borderColor": "#023e8a",
            "borderRadius": 25,
            "borderWidth": 1,
            "display": "flex",
            "flexDirection": "row",
            "height": 150,
            "justifyContent": "space-around",
            "marginLeft": 30,
            "marginTop": 20,
            "width": 400,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": undefined,
            }
          }
          style={
            Object {
              "borderRadius": 50,
              "height": 100,
              "marginTop": 16,
              "width": 100,
            }
          }
          testID="Josep"
        />
        <Text
          style={
            Object {
              "color": "#03045e",
              "fontSize": 18,
              "marginTop": 50,
              "textTransform": "uppercase",
            }
          }
        >
          Josep
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#caf0f8",
            "borderColor": "#023e8a",
            "borderRadius": 25,
            "borderWidth": 1,
            "display": "flex",
            "flexDirection": "row",
            "height": 150,
            "justifyContent": "space-around",
            "marginLeft": 30,
            "marginTop": 20,
            "width": 400,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": undefined,
            }
          }
          style={
            Object {
              "borderRadius": 50,
              "height": 100,
              "marginTop": 16,
              "width": 100,
            }
          }
          testID="Juanito"
        />
        <Text
          style={
            Object {
              "color": "#03045e",
              "fontSize": 18,
              "marginTop": 50,
              "textTransform": "uppercase",
            }
          }
        >
          Juanito
        </Text>
      </View>
      <Text
        style={
          Object {
            "color": "#03045e",
            "fontSize": 12,
            "marginLeft": 280,
            "marginTop": -5,
            "position": "absolute",
          }
        }
      >
        Se han encontrado 
        2
         resultados
      </Text>
    </View>
  </RCTScrollView>
</View>
`;
